[tuf] Store rkth/sign hashes in TUF repo description

karencfv opened the pull request:
This commit introduces a new field in `TufArtifactMeta` that maps an RoT/RoT bootloader artifact with its associated sign. This will be necessary for the planner to support [RoT](https://github.com/oxidecomputer/omicron/pull/8421) and [RoT bootloader](https://github.com/oxidecomputer/omicron/pull/8664) updates.

~NOTE: There is still a test failing. I'm going to investigate what's going on, but I'd like to set this PR as ready to review to get comments on the structure of the tables, etc.~

iliana commented on common/src/api/external/mod.rs:
```diff
@@ -3357,13 +3357,22 @@ pub struct TufRepoDescription {
     /// Information about the artifacts present in the repository.
     pub artifacts: Vec<TufArtifactMeta>,
+
+    // Association between an artifact and its RoT signing information.
+    // This also covers the RoT bootloader artifacts which are signed as well.
+    pub rots_by_sign: Vec<TufRotBySign>,
```
This makes sense for the db-model `TufRepoDescription` but I think we want to do something else for the external API type here. It might be difficult to come up with the right sort of mapping before we implement the planner, but as written I think the planner would need to decide to find an artifact that matches a specific artifact kind and RTKH, and in order to do that it would first need to go through all of `artifacts` to find the artifact IDs matching the kind we want, and then find which artifact ID has the right RTKH by searching `rots_by_sign`.

This makes me lean toward having a `signed_by_rtkh: Option<Vec<u8>>` on the external `TufArtifactMeta`. This is not something we have to do in the db-model structs; we can fill in the `signed_by_rtkh` field when we convert from the db-model crates to the external crates in the `into_external` function.

iliana commented on common/src/api/external/mod.rs:
```diff
@@ -3357,13 +3357,22 @@ pub struct TufRepoDescription {
+    pub rots_by_sign: Vec<TufRotBySign>,
 }
 
 impl TufRepoDescription {
     /// Sorts the artifacts so that descriptions can be compared.
     pub fn sort_artifacts(&mut self) {
         self.artifacts.sort_by(|a, b| a.id.cmp(&b.id));
     }
+
+    /// Sorts the RoTs by sign so that descriptions can be compared.
+    pub fn sort_rots_by_sign(&mut self) {
+        self.rots_by_sign.sort_by(|a, b| a.id.cmp(&b.id));
+    }
 }
```
If we do keep a separate field, we should probably have a single sort method here. Would there be a case where you want to sort one but not the other?

iliana commented on schema/crdb/add-tuf-rot-by-sign/up1.sql:
```diff
@@ -0,0 +1,9 @@
+    name STRING(63) NOT NULL,
+    version STRING(63) NOT NULL,
+    kind STRING(63) NOT NULL,
+    sign Bytes NOT NULL,
```
```suggestion
    sign BYTES NOT NULL,
```

iliana commented on schema/crdb/dbinit.sql:
```diff
@@ -2534,6 +2534,31 @@ CREATE TABLE IF NOT EXISTS omicron.public.tuf_repo_artifact (
     PRIMARY KEY (tuf_repo_id, tuf_artifact_id)
 );
 
+-- Describes an individual association between an RoT or RoT bootloader artifact
+-- and its sign (RKTH) value from an uploaded TUF repo.
+CREATE TABLE IF NOT EXISTS omicron.public.tuf_rot_by_sign (
+    id UUID PRIMARY KEY,
+    name STRING(63) NOT NULL,
+    version STRING(63) NOT NULL,
+    -- The kind of artifact this is
+    kind STRING(63) NOT NULL,
+    -- The sign (RKTH) value associated with this image
+    sign Bytes NOT NUll,
+    -- The time this rot_by_sign was first recorded
+    time_created TIMESTAMPTZ NOT NULL,
+    CONSTRAINT unique_name_version_kind_sign UNIQUE (name, version, kind, sign)
+);
+
+-- Reflects that a particular RoT or RoT bootloader artifact was provided by a
+-- particular TUF repo.
+-- This is a many-many mapping.
+CREATE TABLE IF NOT EXISTS omicron.public.tuf_repo_rot_by_sign (
+    tuf_repo_id UUID NOT NULL,
+    tuf_rot_by_sign_id UUID NOT NULL,
+
+    PRIMARY KEY (tuf_repo_id, tuf_rot_by_sign_id)
+);
```
It feels somewhat dubious to have a separate table with _separate_ "ROT by sign IDs" (which are presumably different from the IDs in the `tuf_artifact` table) that is describing optional data on a TUF artifact. Given the UNIQUE constraint I'm tempted to suggest that this should be a nullable column on the main `tuf_artifact` table.

plotnick commented on common/src/api/external/mod.rs:
```diff
@@ -3408,6 +3408,10 @@ pub struct TufArtifactMeta {
     /// The size of the artifact in bytes.
     pub size: u64,
+
+    /// Sign (root key hash table) hash of a signed image. Only applicable to
+    /// RoT or RoT bootloader artifacts.
+    pub rot_sign: Option<Vec<u8>>,
```
```suggestion
    /// Contents of the `SIGN` field of a Hubris archive caboose, i.e.,
    /// an identifier for the set of valid signing keys. Currently only
    /// applicable to RoT image and bootloader artifacts, where it will
    /// be an LPC55 Root Key Table Hash (RKTH).
    pub sign: Option<Vec<u8>>,
```

plotnick commented on update-common/src/artifacts/update_plan.rs:
```diff
@@ -419,11 +420,13 @@ impl<'a> UpdatePlanBuilder<'a> {
         let (artifact_id, bootloader_caboose) =
             read_hubris_caboose_from_archive(artifact_id, data.clone())?;
 
+        let sign = bootloader_caboose.sign.expect("required SIGN in caboose");
```
I think we should return an appropriate error here instead of panicking. Development images may be missing their `SIGN` fields, and that shouldn't be a fatal error.

plotnick commented on update-common/src/artifacts/update_plan.rs:
```diff
@@ -583,10 +586,9 @@ impl<'a> UpdatePlanBuilder<'a> {
-            kind: artifact_kind,
-            sign: image_b_caboose.sign.expect("required SIGN in caboose"),
-        };
+        let sign_b = image_b_caboose.sign.expect("required SIGN in caboose");
```
Same as above, we shouldn't panic here.

plotnick reviewed (commented):
I think the mechanism here is fine, and the implementation straightforward enough. My main comment is about the terminology, and so the field names.

The `SIGN` field of a Hubris archive intended to run on a RoT contains an LPC55 RKTH value. But those are separate concepts, and the caboose is independent of RoT/SP. I think the only real advantage of using the name `sign` over `rkth` is that it allows for the future possibility of signed SP images (which won't use a RKTH, they'll use something else). To support that future without coming back and renaming everything here, I think we should call this field just `sign`, and similarly drop the `rot_` prefix from the names of all associated database tables, columns, fields, etc.

If we don't want to go in that direction, then I think we should call this field `rot_rkth`. But I think `rot_sign` isn't ideal.

karencfv commented:
Thanks for taking a look @plotnick! I've addressed your comments in [bc394ba](https://github.com/oxidecomputer/omicron/pull/8729/commits/bc394ba4574b42799c69c651cc4d0211be14aaa8)

plotnick reviewed (approved):
Thanks, looks great. Looks like it needs a `cargo xtask openapi generate` update, but otherwise I say :ship:
